### PR TITLE
Migrating Kubebuilder checks to Webhooks

### PR DESCRIPTION
**Pull request by akshmish**
**What this PR does / why we need it**:

Integrates Webhooks with the current NDB Operator functionality, adds validation and defaulter logic for the DB provisioning specs. 

**How Has This Been Tested?**:
Manually tested via deploying the operator in the local K8s cluster.
Tested various combinations of input to make sure expected error is thrown or the request succeeding.

### CODE REVIEW

File: common/constants.go
Comment (manavrajvanshi) [inline common/constants.go:28]: Since these won't be used by anything other than the webhooks in the api/v1alpha1 package, can we move these out to another file or to the api package?

File: api/v1alpha1/database_webhook.go
Comment (manavrajvanshi) [inline api/v1alpha1/database_webhook.go:45]: This might not be required as this being a boolean will automatically be initialized to false.

File: api/v1alpha1/database_webhook.go
Comment (manavrajvanshi) [inline api/v1alpha1/database_webhook.go:112]: Can we remove this validation since the defaulter will make sure that this value is never empty?

File: api/v1alpha1/database_webhook.go
Comment (manavrajvanshi) [inline api/v1alpha1/database_webhook.go:116]: Can we remove this validation since the defaulter will add default elements to this slice?

File: api/v1alpha1/database_webhook.go
Comment (manavrajvanshi) [inline api/v1alpha1/database_webhook.go:145]: Should we only log these errors when there are errors? Thinking about a len check of allErrs in an `if` block that contains line 143-144

File: api/v1alpha1/database_webhook.go
Comment (manavrajvanshi) [inline api/v1alpha1/database_webhook.go:139]: Can/should we change this to mixedCaps instead of snake_case to follow the golang conventions?

File: api/v1alpha1/database_webhook.go
Comment (manavrajvanshi) [inline api/v1alpha1/database_webhook.go:75]: Can/should we change this to mixedCaps instead of snake_case to follow the golang conventions?

**Review by manavrajvanshi — CHANGES_REQUESTED** [review]
Few comments in addition to the offline discussion, LMK if we should discuss more.

File: config/manager/kustomization.yaml
Comment (akshmish) [reply config/manager/kustomization.yaml:15]: Yes, I will make a note to revert this change before merging because I need to revert it locally before every commit as of now.

File: config/crd/patches/cainjection_in_databases.yaml
Comment (akshmish) [reply config/crd/patches/cainjection_in_databases.yaml:6]: added by the Operator SDK.

File: main.go
Comment (akshmish) [reply main.go:109]: Nice idea, done.

File: config/crd/bases/ndb.nutanix.com_databases.yaml
Comment (yash-ntnx) [inline config/crd/bases/ndb.nutanix.com_databases.yaml:87]: I am most likely reading this wrong ... but do we need both id and name? I thought any 1 was sufficient

File: api/v1alpha1/database_webhook.go
Comment (manavrajvanshi) [reply api/v1alpha1/database_webhook.go:53]: Sure that makes sense! We can also make this auto generated like pod names, I think you shared this idea initially too.

Comment (akshmish) [reply api/v1alpha1/database_webhook.go:53]: Yes that is a possibility. What you are saying makes sense to me. What I feel about this is - as long as we are throwing an actionable, meaningful error to the user saying "database instance with the same name exists", this should be acceptable. Let me know your thoughts on it.

File: api/v1alpha1/database_webhook.go
Comment (akshmish) [reply api/v1alpha1/database_webhook.go:160]: If there are no errors, let it print empty list, that way we can log the exact set of responses we received from the webhooks. Since we are not storing the logs in a file, it is just an extra stdout log, probably not going to be as bad.

File: api/webhooks_constants.go
Comment (akshmish) [reply api/webhooks_constants.go:14]: Yes, not as of now, but I am adding since we will add it in future soon. But I am ok to remove it if you still think it shouldn't be there.

File: Makefile
Comment (akshmish) [reply Makefile:125]: So basically for local-dev, we will be running the Operator as a separate go service using "go run ./main.go". We do not want to run webhook in that setting. But webhook will be applied when you are running the operator inside the k8s cluster - and this is how the customers will use NDB operator. So to answer your question in short, yes this is a conscious choice as @manavrajvanshi pointed out.

File: Makefile
Comment (akshmish) [reply Makefile:163]: Good point. I think the YAML at given URL (containing version) still should be available for the backward compatibility purposes. 

I can specify the version inside a constant but I would still prefer to retain an explicit version (for which we are qualifying the NDB operator). Any upgrades in cert-manager versions should be explicit. 

If the version goes E0L, we will get an explicit error and then we can upgrade.

File: api/v1alpha1/database_webhook.go
Comment (akshmish) [reply api/v1alpha1/database_webhook.go:59]: I see your point but I will hold this discussion for the another uber item that I have added in the task sheet. It is "Audit of the required fields and the default values". 

That itself won't be a big change but it will still need a discussion and some consensus.

File: api/v1alpha1/database_webhook.go
Comment (akshmish) [reply api/v1alpha1/database_webhook.go:63]: Same as above. Separate task in the sheet called "Audit of the required fields and the default values"

File: config/crd/bases/ndb.nutanix.com_databases.yaml
Comment (akshmish) [reply config/crd/bases/ndb.nutanix.com_databases.yaml:87]: Turns out that, it was added because I removed "//+optional" from the struct definition. Adding it back remove these.

File: Makefile
Comment (akshmish) [reply Makefile]: I see your point, so I was doing a little bit of research on it. 

As per the [documentation](https://www.gnu.org/software/make/manual/html_node/Rule-Syntax.html), the make target syntax is: 

```
targets : prerequisites
        recipe
        …

```

which means, "prerequisites" go after the ":".


In this case, "force-pull-deploy" target depends on the "manifests" and "kustomize" targets.
So I think we should retain it as it is a pre-requisite regardless of what other target are we running inside the recipe. Besides running those pre-requisites twice may not cause any issues in this case.

File: api/v1alpha1/database_webhook.go
Comment (akshmish) [reply api/v1alpha1/database_webhook.go]: This can be discussed as a part of the additional line item that I have added in the operator task sheet. The default values or mandatory fields need some discussion so I will close it here.

File: Makefile
Comment (akshmish) [reply Makefile]: @manavrajvanshi had a similar comment, replied in his thread inline. We can continue that discussion on that thread.

File: Makefile
Comment (akshmish) [reply Makefile]: Decided to remove the target itself as running the target is not sufficient. We need to wait until all the pods are up inside k8s cluster before running the next "make deploy" command. So I will add a note to mention the pre-requisite steps in README, instead of relying on this target.

File: api/webhooks_constants.go
Comment (akshmish) [reply api/webhooks_constants.go]: Removed it now.

File: Makefile
Comment (akshmish) [reply Makefile]: Removed the target as per the offline discussion.

File: config/crd/bases/ndb.nutanix.com_databases.yaml
Comment (akshmish) [reply config/crd/bases/ndb.nutanix.com_databases.yaml]: As per our offline discussion, if we set it to False, operator will not be able to successfully provision a database right? 
I think once I set it to false, I started seeing X509 cert errors in the controller logs. 

Let's cover the default as a part of the line item, what say?

File: Makefile
Comment (yash-ntnx) [reply Makefile:40]: i'll unresolve this conv. so we remember to back out this change. once backed out, you can resolve it.

File: api/v1alpha1/database_webhook.go
Comment (yash-ntnx) [reply api/v1alpha1/database_webhook.go:48]: Akshay we need to do better than just erroring out. NDB can't "assume" something and then throw an error based on that. If we are assuming, we have to assume more smartly.

File: Makefile
Comment (akshmish) [reply Makefile:40]: Sure, actually I closed it since exactly same comment has been added by @manavrajvanshi, I had left that thread open to avoid duplication. I will close both threads together. 

https://github.com/nutanix-cloud-native/ndb-operator/pull/121#discussion_r1254770973

File: api/v1alpha1/database_webhook.go
Comment (akshmish) [reply api/v1alpha1/database_webhook.go:48]: Sure Yash, the other option I have suggested above is to prefix/suffix timestamp (in a human readable format) to the default Name. What are your thoughts on it?

**Review by manavrajvanshi — CHANGES_REQUESTED** [review]

File: common/util/webhook_utils.go
Comment (akshmish) [reply common/util/webhook_utils.go:44]: As discussed offline, I will rename the method to "IsFeatureEnabled" to make it generic and reusable.

File: api/v1alpha1/database_webhook.go
Comment (akshmish) [reply api/v1alpha1/database_webhook.go:48]: Picking something like this for readability: 

database_2006-Jan-02_03-04-05PM

File: Makefile
Comment (akshmish) [reply Makefile:132]: "export" keyword is not necessary as we are not executing subsequent shell commands in the same shell session. In short, it is not a bad idea to add it but it is not necessary to do so.

File: common/util/webhook_utils.go
Comment (akshmish) [reply common/util/webhook_utils.go:47]: I wanted to keep this method self-sufficient. This method should return the default value instead of the caller catching it and deciding what to do with it. This implementation keeps the client logic much simpler. I can log it's value in the caller code, but I would still prefer to keep the implementation details inside the util method. What do you think?

File: main.go
Comment (akshmish) [reply main.go]: Discussed offline. Renamed the util method.

File: api/v1alpha1/database_webhook.go
Comment (akshmish) [reply api/v1alpha1/database_webhook.go]: As of now, there are certain db spec fields that we are internally populating while generating the provisioning request body, and these fields are not exposed for the user-input as of now.
eg Database Server VM Name, DB description etc

Please refer to db_helpers.go#GenerateProvisioningRequest for the complete list of such fields.

As discussed offline, 
1. As of now, the DB Server VM name is an internally populated field while generating the request body, we are not changing it in this PR. That logic needs to be revisited and a line item has been created in the task sheet.

2. Duplicate database names are ok and will not cause any issues. So we will not change it.

3. Database Instance Name will be made a mandatory field.

Comment (akshmish) [reply api/v1alpha1/database_webhook.go]: Resolving this comment.

File: common/util/webhook_utils.go
Comment (akshmish) [reply common/util/webhook_utils.go:47]: Let me know if it makes sense, resolving the comment, feel free to reopen if you need to.

File: config/crd/bases/ndb.nutanix.com_databases.yaml
Comment (akshmish) [reply config/crd/bases/ndb.nutanix.com_databases.yaml]: I removed the defaulter logic for it, which means the default will be considered as False. Hope that answers your question. Reopen if you have more Qs.

File: common/util/webhook_utils.go
Comment (akshmish) [reply common/util/webhook_utils.go:40]: I believe it can be used to combine multiple errors when you have individual errors as shown in the example:

	err1 := errors.New("err1")
	err2 := errors.New("err2")
	err := errors.Join(err1, err2)
	
	
We have an array of errors here which we are trying to append together, which means we will need to loop through the allErrs anyway, basically voiding the intention to use Join().

Comment (akshmish) [reply common/util/webhook_utils.go:40]: I am closing this comment if it answers your question, feel free to reopen if you have further questions.

File: config/manager/kustomization.yaml
Comment (akshmish) [reply config/manager/kustomization.yaml]: Reverted this change.

File: Makefile
Comment (akshmish) [reply Makefile]: Reverted this change.

**Comment by akshmish** [comment]
I am leaving one Kubebuilder check (	// +kubebuilder:default:="{emptytminfo}") for the next PR as it previously required some hacking and I need to make sure that is not breaking anything.

File: api/v1alpha1/database_types.go
Comment (manavrajvanshi) [reply api/v1alpha1/database_types.go:79]: Cool, but marking it as an optional here and then requiring it in the webhook are two opposite constructs which might lead to confusion for us (the developers). Can/should we remove this annotation or mark it as required again?

File: config/crd/bases/ndb.nutanix.com_databases.yaml
Comment (yash-ntnx) [inline config/crd/bases/ndb.nutanix.com_databases.yaml:108]: is this description accurate? are we still defaulting to 10?

**Review by yash-ntnx — APPROVED** [review]

File: api/v1alpha1/database_webhook.go
Comment (manavrajvanshi) [inline api/v1alpha1/database_webhook.go]: How can we also handle the case when the databaseInstanceName is empty?

File: api/v1alpha1/database_webhook.go
Comment (manavrajvanshi) [inline api/v1alpha1/database_webhook.go]: How can we also handle the case when the databaseInstanceName is empty?

File: api/v1alpha1/database_webhook.go
Comment (manavrajvanshi) [reply api/v1alpha1/database_webhook.go:188]: Unresolving this comment, looks like this was missed in the commits.

File: config/crd/bases/ndb.nutanix.com_databases.yaml
Comment (akshmish) [reply config/crd/bases/ndb.nutanix.com_databases.yaml]: Good catch, no I removed the defaulter logic and retained the original validation logic for the minimum 10 GBs value. Fixed it in this PR itself.

File: common/util/webhook_utils.go
Comment (manavrajvanshi) [inline common/util/webhook_utils.go:49]: NIT: We can update the comment to say `IsFeatureEnabled` instead of `IsWebhiikDisabled`.

File: config/crd/bases/ndb.nutanix.com_databases.yaml
Comment (manavrajvanshi) [inline config/crd/bases/ndb.nutanix.com_databases.yaml:59]: Since we're not defaulting anymore, can we update this? 
Maybe try running make generate manifests to see if this gets auto generated?

File: config/crd/bases/ndb.nutanix.com_databases.yaml
Comment (manavrajvanshi) [inline config/crd/bases/ndb.nutanix.com_databases.yaml:133]: NIT: Uppercase for NONE

File: config/crd/bases/ndb.nutanix.com_databases.yaml
Comment (manavrajvanshi) [inline config/crd/bases/ndb.nutanix.com_databases.yaml:166]: Can we remove the default since it is a required field now?

**Review by manavrajvanshi — CHANGES_REQUESTED** [review]
Minor comments, rest LGTM!

File: api/v1alpha1/database_webhook.go
Comment (akshmish) [reply api/v1alpha1/database_webhook.go:188]: Done.

**Review by manavrajvanshi — APPROVED** [review]

**Review by manavrajvanshi — CHANGES_REQUESTED** [review]
Can you also add the e2e test for this? Thanks!

File: api/v1alpha1/database_webhook.go
Comment (akshmish) [reply api/v1alpha1/database_webhook.go]: So the request will not pass through the Validator if the DatabaseInstanceName is empty (DatabaseInstanceName validation down the line), so the TM name will not be applied actually. The Webhook flow naturally takes care of it instead of adding in-place additional checks.

File: config/crd/bases/ndb.nutanix.com_databases.yaml
Comment (akshmish) [reply config/crd/bases/ndb.nutanix.com_databases.yaml]: Yes this gets auto generated. running make generate fixed this. I think we can make a single target to be run before commiting to git. eg make commit, which will run all targets in one run.

File: api/v1alpha1/database_webhook.go
Comment (akshmish) [reply api/v1alpha1/database_webhook.go]: We are not defaulting to any name for DatabaseInstanceName so the webhook flow will take care of it. The validation webhook will fail down the line if DatabaseInstanceName is empty.

**Comment by akshmish** [comment]
> Can you also add the e2e test for this? Thanks!

Yes, I had this discussion with @yash-ntnx. I believe you have added a few E2E tests, which I need to understand how they work. I need to reuse the same test, just create multiple inputs to simulate the failure, but it will also involve running the Operator inside K8s cluster, so need some research on how our testing framework simulates operations.

File: api/v1alpha1/database_types.go
Comment (manavrajvanshi) [inline api/v1alpha1/database_types.go:78]: Shouldn't we remove the optional since it is required?

**Review by manavrajvanshi — APPROVED** [review]
LGTM! But still confused about server being optional and required at the same time.
Regarding the e2e test - I meant manual test of some scenarios until the automation is merged..

**Pull request merged**
